internal/providers/github/github.go

```diff
@@ -115,3 +126,78 @@ func ParseV1Config(rawCfg json.RawMessage) (*minderv1.GitHubProviderConfig, erro
 
 	return w.GitHub, nil
 }
+
+func (c *RestClient) waitIfRateLimited(ctx context.Context) error {
+	blockedChan, ok := c.tokenStatStore.GetBlockingChan(c.token, db.ProviderTypeGithub)
+	if !ok {
+		return nil
+	}
+
+	select {
+	case <-blockedChan:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-time.After(MaxRateLimitWait):
+		return fmt.Errorf("rate limit wait exceeded for %s", c.GetOwner())
+	}
+}
+
+func (c *RestClient) checkRateLimitResponse(ctx context.Context, resp *http.Response) error {
+	if resp.StatusCode != http.StatusForbidden && resp.StatusCode != http.StatusTooManyRequests {
+		return nil
+	}
+
+	logger := zerolog.Ctx(ctx)
+
+	// Rate limit checking follows optimistic concurrency model. Race conditions are possible
+	// which can lead to a token being already blocked. In such cases, we do not block again,
+	// but instead wait for the existing block to be cleared.
+	// Optimistic concurrency is used to avoid the overhead of a mutex lock for every request.
+	err := c.waitIfRateLimited(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Check if rate limit headers are present
+	remaining, _ := strconv.Atoi(resp.Header.Get("x-ratelimit-remaining"))
+	resetTimeUnix, _ := strconv.ParseInt(resp.Header.Get("x-ratelimit-reset"), 10, 64)
+	retryAfter, _ := strconv.Atoi(resp.Header.Get("retry-after"))
+
+	if remaining == 0 && resetTimeUnix > 0 {
+		waitTime := time.Until(time.Unix(resetTimeUnix, 0).UTC())
+
+		logger.Warn().Msgf("Primary rate limit exceeded by %s. Waiting for %s", c.GetOwner(), waitTime)
+		return c.blockAndWaitWithTimeout(ctx, waitTime, MaxRateLimitWait)
+	}
+
+	if retryAfter > 0 {
+		waitTime := time.Duration(retryAfter) * time.Second
+
+		logger.Warn().Msgf("Secondary rate limit exceeded by %s. Waiting for %s", c.GetOwner(), waitTime)
+		return c.blockAndWaitWithTimeout(ctx, waitTime, MaxRateLimitWait)
+	}
+
+	logger.Warn().Msgf("Rate limit exceeded by %s. Waiting for 1 minute before retrying", c.GetOwner())
```

Comment:
Could we extract `1*time.Minute` to a var or a const to make sure we don't forget updating the log message if we update the wait time?